### ISSUE TITLE
Error fetching ubuntu vulnerabilies

### ISSUE THREAD

**Issue by chka** [quay/clair#323]
I've tried the kubernetes deployment and I'm getting the following error on the clair pod:

```
2017-02-22 17:07:59.686101 E | updater/fetchers/ubuntu: could not branch Ubuntu repository: exit status 3. output: bzr: ERROR: Connection error: Couldn't resolve host 'launchpad.net' [Errno -2] Name or service not known
2017-02-22 17:07:59.686250 E | updater: an error occured when fetching update 'Ubuntu': could not download requested resource.
```

Any ideas?

**Comment by Quentin-M** [comment]
Hi,

Clearly, you have a DNS issue in your setup. I'd investigate why your Kubernetes Pod doesn't have proper DNS resolution. Have you installed the [DNS Addon](https://coreos.com/kubernetes/docs/latest/deploy-addons.html)?

**Comment by jdel** [comment]
I am experiencing the same issue with clair 1.2.6 running on a coreos vm in azure.

i am running clair with --log-level=debug. Is there any way to get more information on what fails ?

```
2017-02-23 15:25:45.805146 D | updater: attempting to obtain update lock
2017-02-23 15:25:46.608401 I | updater: updating vulnerabilities
2017-02-23 15:25:46.608461 I | updater: fetching vulnerability updates
2017-02-23 15:25:46.608487 I | updater/fetchers/debian: fetching Debian vulnerabilities
2017-02-23 15:25:46.618007 I | updater/fetchers/rhel: fetching Red Hat vulnerabilities
2017-02-23 15:25:46.618099 I | updater/fetchers/ubuntu: fetching Ubuntu vulnerabilities
2017-02-23 15:25:48.351953 D | updater/fetchers/rhel: no Red Hat update.
2017-02-23 15:27:41.628118 E | updater/fetchers/ubuntu: could not branch Ubuntu repository: signal: killed. output:
2017-02-23 15:27:41.634019 E | updater: an error occured when fetching update 'Ubuntu': could not download requested resource.
2017-02-23 15:27:41.634665 I | updater: adding metadata to vulnerabilities
```

**Comment by Quentin-M** [comment]
@jdel Your error is different. Your bzr process got killed. This typically means that you ran out of memory. I'd check your system logs.

**Comment by jdel** [comment]
What kind of range should I look at to make Clair perform as it should ? Currently the container is running with a memory limit of 1GB

**Comment by Quentin-M** [comment]
@jdel Yeah, Clair doesn't require a lot of memory by itself, but bzr (used for Ubuntu vulnerabilities) kinda does, especially during the initial update! I don't have a clear idea as of now but I could read our metrics if you are interested in having proper data. Off the top of my head, we run with 4GB or 8GB. Because we answer thousands of requests per second, we actually have two separate cluster stacks of Clair to support our load. One is strictly for analyzing images (each request hits one CPU core pretty hard due to uncompression, no latency constrait), the other to answer requests (e.g. what are the vulnerabilities of this image, low CPU usage but low latency required).

**Comment by chka** [comment]
I'm confused for 3 reasons:

1. The pod is fine resolving any host (including launchpad.net) without the need for DNS addon 
```
root@clair-0ls07:/go/src/github.com/coreos/clair# ping launchpad.net
PING launchpad.net (91.189.89.222): 56 data bytes
64 bytes from 91.189.89.222: icmp_seq=0 ttl=61 time=21.431 ms
```
2. I've added the DNS addon and I'm still getting the exact same error and only on Ubuntu.

```
2017-02-24 09:22:37.477651 I | updater/fetchers/ubuntu: fetching Ubuntu vulnerabilities
2017-02-24 09:22:37.496727 I | updater/fetchers/debian: fetching Debian vulnerabilities
2017-02-24 09:22:37.497200 I | updater/fetchers/rhel: fetching Red Hat vulnerabilities
2017-02-24 09:22:37.650401 E | updater/fetchers/ubuntu: could not branch Ubuntu repository: exit status 3. output: bzr: ERROR: Connection error: Couldn't resolve host 'launchpad.net' [Errno -2] Name or service not known
2017-02-24 09:22:37.650513 E | updater: an error occured when fetching update 'Ubuntu': could not download requested resource.
```

3. It kind of freezes for approximately 30 minutes in this point and after that, it continues fetching the vulnerabilities. After that, approximately every minute loops around and produces this error:

```
2017-02-24 09:59:58.992700 I | updater: update finished
2017-02-24 09:59:58.995731 I | updater: updating vulnerabilities
2017-02-24 09:59:58.995806 I | updater: fetching vulnerability updates
2017-02-24 09:59:58.995865 I | updater/fetchers/rhel: fetching Red Hat vulnerabilities
2017-02-24 09:59:58.995955 I | updater/fetchers/ubuntu: fetching Ubuntu vulnerabilities
2017-02-24 09:59:59.024887 I | updater/fetchers/debian: fetching Debian vulnerabilities
2017-02-24 09:59:59.161926 E | updater/fetchers/ubuntu: could not pull Ubuntu repository: exit status 3. output: bzr: ERROR: Not a branch: "/tmp/ubuntu-cve-tracker757773204/repository/.bzr/branch/": location is a repository.
2017-02-24 09:59:59.162040 E | updater: an error occured when fetching update 'Ubuntu': could not download requested resource.
2017-02-24 10:00:01.989971 I | updater: adding metadata to vulnerabilities
```
If I check the DB, I'm getting this:
```
postgres=# select count(*) from vulnerability;
 count 
-------
 63891
(1 row)
```



Any ideas?

**Comment by jdel** [comment]
I upgraded to clair:v2.0.0-pre/4 yesterday I have a different error message in the logs now:

```
2017-02-23 16:40:12.880370 E | updater/fetchers/ubuntu: could not pull Ubuntu repository: exit status 3. output: bzr: ERROR: Not a branch: "/tmp/ubuntu-cve-tracker446924345/repository/.bzr/branch/": location is a repository.
2017-02-23 16:40:12.880476 E | updater: an error occured when fetching update 'Ubuntu': could not download requested resource.
```

Might be related ?

**Comment by chka** [comment]
I'm closing this issue as clair works much better standalone and not inside a container

**Comment by jdel** [comment]
Is this an acceptable answer when coreos provides an official clair container ? I don't think so.

Anyway, I am currently trying out clair:v2.0.0-rc-1 container and I still experience the same issue.

Running bzr manually from inside the container shows the same problem, it downloads happily for a while then gets killed for no reason.

I will investigate a bit more later.

**Comment by jdel** [comment]
For posterity, here is what works for me.

It seems that bzr likes a lot of ram when doing the `bzr branch` command to fetch ubuntu vulnerabilities.

So if there is no docker memory limit on the container, the server running the clair container most likely doesn't have enough ram.

I have now set the mem limit to 2Gb and I can pull ubuntu vulnerabilities without any issue.

**Comment by jzelinskie** [comment]
Thanks @jdel I think we could avoid some of this headache in the future if we put RAM recommendations in the README.

**Comment by Quentin-M** [comment]
We have exclusively been running Clair in containers, yes.

Clair itself does not eat lots of memory at all. However, when bzr does its initial update or any large updates, it will eats up to 2-4GB memory. My second answer seems to address the cause.

One approach is to dedicate one deployment to doing updates, and another deployment to be API bees. What we decided to go with in production is to run two deployments of pretty small machines: one that would process images (lots of CPU usage due to tar unarchiving of the images, no hard response time constraint) and run the upgrades, and another one responsible for responding to the API calls for getting vulnerabilities (barely any CPU usage - mostly database roundtrip and waiting, response time constrained).

**EDIT**: OK Actually, I came with the exact same answer in this thread on Feb 24: https://github.com/coreos/clair/issues/323#issuecomment-282243422
